### PR TITLE
ListUsers: vary CDN and memcache cache

### PR DESCRIPTION
**Pull request by macbre**
https://wikia-inc.atlassian.net/browse/SUS-3207

Vary cache entries on user language - cached data has localized messages.

And use `RequestContext` instead of global variables

### CODE REVIEW

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 1]
```diff
@@ -96,7 +96,7 @@ function setOrder    	( $orders = array() ) {
 	function getGroups   	() { return $this->mGroups; }
 
 	public function loadData() {
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:98]: MAJOR The Cyclomatic Complexity of this function "loadData" is 26 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)

[project-id]: wikia:app
MAJOR This function "loadData" has 205 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 2]
```diff
@@ -219,12 +224,12 @@ public function loadData() {
 						0 => "",
 						1 => $sk->makeLinkObj(
 							Title::newFromText( 'Contributions', NS_SPECIAL ),
-							$wgLang->ucfirst( wfMsg('contribslink') ),
+							$lang->ucfirst( wfMsg('contribslink') ),
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:227]: MAJOR Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 3]
```diff
@@ -239,19 +244,19 @@ public function loadData() {
...
-							$wgLang->ucfirst( wfMsg('blocklink') )
+							$lang->ucfirst( wfMsg('blocklink') )
 						);
 					}
-					if ( $wgUser->isAllowed( 'userrights' ) && ( !$userIsBlocked ) ) {
+					if ( $user->isAllowed( 'userrights' ) && ( !$userIsBlocked ) ) {
 						$links[] = $sk->makeLinkObj(
 							Title::newFromText( 'UserRights', NS_SPECIAL ),
-							$wgLang->ucfirst( wfMsg('listgrouprights-rights') ),
+							$lang->ucfirst( wfMsg('listgrouprights-rights') ),
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:259]: MAJOR Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 4]
```diff
@@ -239,19 +244,19 @@ public function loadData() {
...
 							"user={$oEncUserName}"
 						);
 					};
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:262]: MINOR Remove this empty statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1116)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 5]
```diff
@@ -239,19 +244,19 @@ public function loadData() {
 					}
 
 					if ( $oUTitle instanceof Title ) {
-						$links[0] = $sk->makeLinkObj( $oUTitle, $wgLang->ucfirst(wfMsg($msg) ) );
+						$links[0] = $sk->makeLinkObj( $oUTitle, $lang->ucfirst(wfMsg($msg) ) );
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:247]: MAJOR Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 6]
```diff
@@ -219,12 +224,12 @@ public function loadData() {
...
 						1 => $sk->makeLinkObj(
 							Title::newFromText( 'Contributions', NS_SPECIAL ),
-							$wgLang->ucfirst( wfMsg('contribslink') ),
+							$lang->ucfirst( wfMsg('contribslink') ),
 							"target={$oEncUserName}"
 						),
 						2 => $sk->makeLinkObj(
 							Title::newFromText( 'Editcount', NS_SPECIAL ),
-							$wgLang->ucfirst( wfMsg('listusersedits') ),
+							$lang->ucfirst( wfMsg('listusersedits') ),
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:232]: MAJOR Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 7]
```diff
@@ -105,14 +105,19 @@ public function loadData() {
...
 		$orderby = implode(",", $this->mOrder);
 		$subMemkey = array(
 			'G'  . implode(",", is_array($this->mFilterGroup) ? $this->mFilterGroup : array()),
 			'U'  . $this->mUserId,
 			'E'  . $this->mEditsThreshold,
 			'O'  . $this->mOffset,
 			'L'  . $this->mLimit,
-			'O'  . $orderby
+			'O'  . $orderby,
+			'L'  . $lang->getCode(), // localized messages are cached, vary by user language
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:120]: INFO Move this trailing comment on the previous empty line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS139)

[project-id]: wikia:app

File: extensions/wikia/Listusers/SpecialListusers_helper.php [hunk 8]
```diff
@@ -239,19 +244,19 @@ public function loadData() {
...
-						$links[0] = $sk->makeLinkObj( $oUTitle, $wgLang->ucfirst(wfMsg($msg) ) );
+						$links[0] = $sk->makeLinkObj( $oUTitle, $lang->ucfirst(wfMsg($msg) ) );
 					}
 
-					if ( $wgUser->isAllowed( 'block' ) && ( !$userIsBlocked ) ) {
+					if ( $user->isAllowed( 'block' ) && ( !$userIsBlocked ) ) {
 						$links[] = $sk->makeLinkObj(
 							Title::newFromText( "BlockIP/{$oUser->getName()}", NS_SPECIAL ),
-							$wgLang->ucfirst( wfMsg('blocklink') )
+							$lang->ucfirst( wfMsg('blocklink') )
```
Comment (WikiaTech) [inline extensions/wikia/Listusers/SpecialListusers_helper.php:253]: MAJOR Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)

[project-id]: wikia:app

**Comment by WikiaTech** [comment]
SonarQube analysis reported 19 issues
* ![MAJOR][MAJOR] 15 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [SpecialListusers_helper.php#L85](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L85): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. ![MAJOR][MAJOR] [SpecialListusers_helper.php#L142](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L142): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
1. ![MAJOR][MAJOR] [SpecialListusers_helper.php#L216](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L216): Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)
1. ![MAJOR][MAJOR] [SpecialListusers_helper.php#L280](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L280): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
1. ![MAJOR][MAJOR] [SpecialListusers_helper.php#L322](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L322): Replace this legacy wfMsg* function call with one of the new i18n functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3ALegacyWfMsgUsage)
1. ![MAJOR][MAJOR] [table.js#L38](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/js/table.js#L38): Function has a complexity of 13 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. ![MAJOR][MAJOR] [table.js#L51](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/js/table.js#L51): Restrict what this loop acts on by testing each property. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AForIn)
1. ![MAJOR][MAJOR] [table.js#L54](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/js/table.js#L54): Add a "default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3ASwitchWithoutDefault)
1. ![INFO][INFO] [SpecialListusers_helper.php#L469](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L469): Complete the task associated to this "TODO" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1135)
1. ![INFO][INFO] [SpecialListusers_helper.php#L469](https://github.com/Wikia/app/blob/a4d51e9636bac405877f3413e140fcb9de330e3c/extensions/wikia/Listusers/SpecialListusers_helper.php#L469): Move this trailing comment on the previous empty line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS139)

[project-id]: wikia:app
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by Grunny — APPROVED** [review]

**Pull request merged**